File: solutions/ruby/bank-account/1/bank_account.rb
class BankAccount
  def initialize
    @money = 0
    @active = false
  end

  def open
    raise ArgumentError.new('You can\'t open an already open account') if @active
    @active = true
  end

  def close
    raise ArgumentError.new('You can\'t close an already closed account') unless @active
    @active = false
    @money = 0
  end
  
  def deposit(amount)
    raise ArgumentError.new('You can\'t deposit money into a closed account') unless @active
    raise ArgumentError.new('You can\'t deposit a negative amount') unless amount >= 0
    @money += amount
  end

  def withdraw(amount)
    raise ArgumentError.new('You can\'t withdraw money into a closed account') unless @active
    raise ArgumentError.new('You can\'t withdraw more than you have') unless @money >= amount
    raise ArgumentError.new('You can\'t withdraw a negative amount') unless amount >= 0
    @money -= amount
  end

  def balance
    raise ArgumentError.new('You can\'t check the balance of a closed account') unless @active
    @money
  end
  
end

File: solutions/cpp/bank-account/1/bank_account.cpp
#include "bank_account.h"

namespace Bankaccount {

int Bankaccount::balance() const {
    if (m_open) {
        return m_balance;
    }
    throw (std::runtime_error("Account is closed."));
}
void Bankaccount::open() {
    if (!m_open) {
        hands_off.lock();
        m_open = true;
        hands_off.unlock();
    } else {
        throw (std::runtime_error("Account is open."));
    }
}
void Bankaccount::close() {
    if (m_open) {
        hands_off.lock();
        m_open = false;
        m_balance = 0;
        hands_off.unlock();
    } else {
        throw (std::runtime_error("Account is closed."));
    }
}
void Bankaccount::deposit(int amount) {
    if (m_open) {
        if (amount >= 0) {
            hands_off.lock();
            m_balance += amount;
            hands_off.unlock();
        } else {
            throw (std::runtime_error("Amounts cannot be negative."));
        }
    } else {
        throw (std::runtime_error("Account is closed."));
    }
}
void Bankaccount::withdraw(int amount) {
    if (m_open) {
        if (amount >= 0) {
            if (amount <= m_balance) {
                hands_off.lock();
                m_balance -= amount;
                hands_off.unlock();
            } else {
                throw (std::runtime_error("Requested amount not available."));
            }
        } else {
            throw (std::runtime_error("Amounts cannot be negative."));
        }
    } else {
        throw (std::runtime_error("Account is closed."));
    }
}
}
